Integrated Distributed ThinLTO (DTLTO): Design Overview

bd1976bris opened the pull request:
# Initial DTLTO Support

This PR introduces initial support for Integrated Distributed ThinLTO (DTLTO). DTLTO was previously discussed in an [RFC](https://discourse.llvm.org/t/rfc-integrated-distributed-thinlto/69641) and during the LTO roundtable discussion at the October US 2024 LLVM conference. PlayStation has offered this feature as a proprietary technology for some time, and we would like to see support in LLVM.

## Overview of DTLTO

DTLTO enables the distribution of backend ThinLTO compilations via external distribution systems, such as Incredibuild. Existing support for distributing ThinLTO compilations typically involves separate thin-link (`--thinlto-index-only`), backend compilation, and link steps coordinated by a modern build system, like Bazel. This "Bazel-style" distributed ThinLTO requires a modern build system as it must handle the [dynamic dependencies](https://buck2.build/docs/rule_authors/dynamic_dependencies/) specified in the summary index file shards. However, adopting a modern build system can be prohibitive for users with established build infrastructure.

In contrast, DTLTO manages distribution within LLVM during the traditional link step. This approach means that DTLTO is usable with any build process that supports in-process ThinLTO.

## Documentation and Resources

- **RFC**: [Integrated Distributed ThinLTO RFC](https://discourse.llvm.org/t/rfc-integrated-distributed-thinlto/69641)
- **Documentation**:
  - Feature documentation added in this PR at [`llvm/docs/dtlto.rst`](https://github.com/bd1976bris/llvm-project/blob/main/llvm/docs/DTLTO.rst).
  - Clang and LLD CLI documentation added at [`clang/docs/dtlto.rst`](https://github.com/bd1976bris/llvm-project/blob/main/clang/docs/ThinLTO.rst) and [`lld/docs/dtlto.rst`](https://github.com/bd1976bris/llvm-project/blob/main/lld/docs/DTLTO.rst).
- **Example Distributors**: Included distributors for LIT testing added under `llvm/utils/dtlto`.
- **Real Distribution System Support**: Example scripts which work with real distribution systems are available for [Windows](https://gist.github.com/bd1976bris/d142fe0e656ac92dbaa940f77c6dbfa5) and [Linux](https://gist.github.com/nga888/3879816706c803d0c0087ff8bb662cd4#reclient-llvm-dtlto-distributor).

## Features of This Initial Commit

This commit provides a minimal but functional implementation of DTLTO, which will be expanded in subsequent commits. The current implementation includes:

- COFF and ELF support.
- Support for bitcode members in thin archives.
- Basic support for distributing backend ThinLTO compilations.
- A JSON interface that allows new distribution systems to be supported without modifying LLVM.

The goal of this initial commit is to demonstrate what will be required to support DTLTO while providing useful minimal functionality. Hopefully, having a concrete PR will facilitate discussion and review of the feature.

## Performance

We have access to a large farm of computers on Windows. For a link of `clang.exe` on a modest Windows development machine (AMD64 16 cores, 64GB RAM) DTLTO (via [`sn-dbs.py`](https://gist.github.com/bd1976bris/d142fe0e656ac92dbaa940f77c6dbfa5)) was approximately 4 times as fast as multi-threaded in-process ThinLTO.

To estimate the overhead from DTLTO vs in-process ThinLTO, we measured the difference in the time taken to link `Clang` with in-process ThinLTO using one thread per core, and DTLTO using one local process per core. On both Windows and Linux the overhead was approximately 6%.

Note that, to facilitate review, this PR elides performance optimizations where possible.

## Planned Future Enhancements

The following features will be addressed in future commits:

- Support for the ThinLTO cache.
- Support for (non-thin) archives/libraries containing bitcode members.
- Support for more LTO configuration states e.g., basic block sections.
- Performance improvements. For example, improving the performance of the temporary file removal.

## Discussion Points

- **Feature Name**: The `DTLTO` name could potentially cause confusion with the existing Bazel-style distributed ThinLTO. At the LLVM roundtable discussion no one objected to the name, but we remain open to suggestions.
- **Backend Compilation Configuration**: Currently, `Clang` is invoked to do the backend compilations and a minimal number of options are added to the `Clang` command line to ensure that the codegen is reasonable (for the testing we have done so far). However, it would be good to find a scalable solution for matching the code-generation state in the invoked external tool to the code-generation state if an in-process ThinLTO backend was in use.
- **Clang Error Handling**: There is some precedent for compiler drivers handling options that only apply to specific linkers. Should Clang emit an error if DTLTO options are used and the linker isn't LLD?

## Other approaches

We have experimented with other approaches for implementing DTLTO. In particular we have explored:

* Not using a new ThinLTO backend.
* Various ways to handle (non-thin) archives.
* Use of dynamic library plugins instead of processes.

We have prepared another branch to demonstrate some of these ideas: [integrated-DTLTO-no-backend](https://github.com/romanova-ekaterina/llvm-project/pull/new/kromanova/main/integrated-DTLTO-no-backend)

## List of Child PRs:
_(I intend to update this section as new PRs are filed.)_

**LLVM 21:**
- ✅ **Core LLVM functionality** (merged)
  - PR #127749 
- ✅ **ELF LLD support: Note no thin archive support** (merged)
  -  PR #142757
- ✅ **LLD ELF Support bitcode members of thin archives** (merged)
  - PR #149425
- ✅ **Clang UI** (merged)
  - PR #147265
- ✅ **COFF LLD support** (merged)
  - PR #147265
- ✅ **COFF LLD make /wholearchive work with thin archives** (merged)
  - PR #145487 (merged)

**LLVM 22:**
- ✅ **DTLTO in-process ThinLTO cache support**
  - PR #156433 (merged)
- ✅ **non-thin archive support for DTLTO**
  - PR #157043 (merged)
- ✅ **Support for multi-call LLVM driver toolchain**
  - PR #162456 (merged)
- ✅ **Add Time Trace Scopes for the DTLTO ThinLTO backend**
  - PR #171600 (merged)
- ✅ **Remove temporary files on abnormal exit, e.g. CTRL-C**
  - PR #172280 (merged)
- 🔄 **Normalize paths to avoid 8.3 short paths**
  - PR #175196 (In review)

bd1976bris committed:
Implement integrated distribution for ThinLTO (DTLTO). ELF and COFF only.

bd1976bris commented:
> ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> 
> You can test this locally with the following command:
> View the diff from clang-format here.

I believe this failure is OK as I have followed the (non-standard) formatting in the flagged file which the code for the other ThinLTO backends use.

tru commented on clang/include/clang/Driver/Options.td:
```diff
@@ -969,6 +969,10 @@ def Xlinker : Separate<["-"], "Xlinker">, Flags<[LinkerInput, RenderAsInput]>,
   Visibility<[ClangOption, CLOption, FlangOption]>,
   HelpText<"Pass <arg> to the linker">, MetaVarName<"<arg>">,
   Group<Link_Group>;
+def Xdist : Separate<["-"], "Xdist">, Flags<[LinkOption]>,
```
Not sure I am a fan of Xdist here. Since the feature is called DTLTO why not Xdtlo, dist is a bit ambiguous.

bd1976bris replied:
Good comment, thanks. In this implementation, there are two things to which a user might want to pass arguments:

1. A distributor – which is whatever process is invoked to consume the JSON and distribute the ThinLTO backend compilations.
2. Whatever tool will be invoked by the distributors to perform the ThinLTO backend compilations – in this PR, that is currently `clang`; however, we refer to this as the `remote optimization tool` since it is possible to swap to a different tool without changing the distributors.

We went with `-Xdist` as it indicates that the arguments will be forwarded to the distributor process and is short (to reduce verbosity if there are many forwarded arguments). What would you think about `-Xthinlto-distributor=` instead?

ilovepi replied:
I think being explicit is preferred for compiler options.

MaskRay replied:
Agreed that `-Xdist` is confusing. -Xthinlto-distributor looks good. If someone can come with up with a shorter, non-confusing one, that'd be nice:)

bd1976bris replied:
I'll wait another day and then change to `-Xthinlto-distributor` if there isn't a better suggestion :)

bd1976bris replied:
I have updated to use `-Xthinlto-distributor`. It's longer than `-Xdist` so I have used `CommaJoined`  which allows e.g. `-fthinlto-distributor=incredibuild.exe -Xthinlto-distributor=--verbose,--j10`

tru commented on lld/COFF/LTO.cpp:
```diff
@@ -116,7 +117,18 @@ BitcodeCompiler::BitcodeCompiler(COFFLinkerContext &c) : ctx(c) {
+    StringRef version = getenv("LLD_VERSION"); // For testing only.
+    if (version.empty())
+      version = ctx.saver.save(getLLDVersion());
+    backend = lto::createOutOfProcessThinBackend(
```
When I read this code at first I was confused why it needed to have the arguments for the parallelization, I looked at the comment in LTO.h later and realized that the arguments where not used for the actual codegen, but for generating the index. But I wonder if we even need to pass that as an argument then - it reads to me like it's controlling the external process in some way.

bd1976bris replied:
Thanks. It may be unnecessary. https://github.com/llvm/llvm-project/commit/2edd897a4227e481af33e8e43090ab088cd9d953 implies that the summary index and import list generation can be expensive though which is why this setting is provided. Also some distribution systems have a mechanism to limit how many processes they launch concurrently on the local machine - we could pass the parallelization level to the distributors (by adding it to the JSON) so that this setting may be applied by the distributor?

tru reviewed (commented):
Hello! I had a quick glance at this and while I am no expert in the actual LTO internal code, I was totally able to understand how we could integrate this with our build system (fastbuild) so I am really happy you have decided to upstream this. I am planning of doing an internal prototype of this soon to understand how it would affect our applications.

That said - I hope that someone that's more qualified in the code to review the actual implementation.

tru commented:
> > ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> > You can test this locally with the following command:
> > View the diff from clang-format here.
> 
> I believe this failure is OK as I have followed the (non-standard) formatting in the flagged file which the code for the other ThinLTO backends use.

I agree.

bd1976bris commented:
> Hello! I had a quick glance at this and while I am no expert in the actual LTO internal code, I was totally able to understand how we could integrate this with our build system (fastbuild) so I am really happy you have decided to upstream this. I am planning of doing an internal prototype of this soon to understand how it would affect our applications.

Great. Actually, we wanted to add support for distributing via FASTBuild, icecream and distcc with the initial DTLTO PR. However, we investigated found that these couldn't be supported without extending the existing codebases. More details to follow..

ilovepi commented on clang/docs/ThinLTO.rst:
```diff
@@ -240,6 +240,38 @@ The ``BOOTSTRAP_LLVM_ENABLE_LTO=Thin`` will enable ThinLTO for stage 2 and
+
+If ``-fthinlto-distributor=`` is specified Clang supplies the path to a
+distributable optimization and code generation tool to LLD. Currently this tool
+is Clang itself specified.
```
Typo?
```suggestion
is Clang itself.
```

tru commented:
> Great. Actually, we wanted to add support for distributing via FASTBuild, icecream and distcc with the initial DTLTO PR. However, we investigated found that these couldn't be supported without extending the existing codebases. More details to follow..

Interesting. Interested in hearing more. We have a heavily modified version of FASTbuild and I am very comfortable in that code base - so happy to help if needed.

ilovepi commented on lld/ELF/InputFiles.cpp:
```diff
@@ -1702,6 +1703,38 @@ static uint8_t getOsAbi(const Triple &t) {
+}
+
+// Compute a thin archive member full file path.
+std::string computeFullThinArchiveMemberPath(const StringRef modulePath,
```
Given that we have "Full LTO" and "Thin LTO", I think you may want to rename this to something like `computeThinArchiveMemberFullPath`, to avoid confusion. My first reading of this had me assuming it would handle both Full and Thin LTO somehow before reading the comment.

bd1976bris replied:
Thanks. I have renamed as you suggested.

NuriAmari commented on llvm/lib/LTO/LTO.cpp:
```diff
@@ -2142,3 +2190,327 @@ std::vector<int> lto::generateModulesOrdering(ArrayRef<BitcodeModule *> R) {
+    return Error::success();
+  }
+
+  // Derive a set of Clang options that will be shared/common for all DTLTO
```
With or without this patch to add integrated dthin-lto, I think something like this (automatic translation of a thin-link invocation into a matching Clang backend invocation) would be useful. Optimization level and flags forwarded through Clang directly to the linker (ex: `-Wl,-mllvm,-my_codegen_option`) are those I've run into most often.

For Buck we either have this [imho unfortunate script wrapping the backend Clang invocations](https://github.com/facebook/buck2/blob/main/prelude/cxx/dist_lto/tools/dist_lto_opt_gnu.py) that translates args from the thin-link invocation to the Clang backend invocation, or we specify both sets of flags manually. @teresajohnson @MaskRay How does Bazel handle this problem?

bd1976bris replied:
> With or without this patch to add integrated dthin-lto, I think something like this (automatic translation of a thin-link invocation into a matching Clang backend invocation) would be useful. Optimization level and flags forwarded through Clang directly to the linker (ex: `-Wl,-mllvm,-my_codegen_option`) are those I've run into most often.
> 
> For Buck we either have this [imho unfortunate script wrapping the backend Clang invocations](https://github.com/facebook/buck2/blob/main/prelude/cxx/dist_lto/tools/dist_lto_opt_gnu.py) that translates args from the thin-link invocation to the Clang backend invocation, or we specify both sets of flags manually. @teresajohnson @MaskRay How does Bazel handle this problem?

Thanks for this comment and for (possibly reluctantly) showing the `buck2` script :). This issue has been the most difficult problem in trying to work out how to upstream this. What we have included in this PR is minimal (although functional enough for simple scenarios).

For our downstream proprietary toolchain, it is relatively easy to translate LTO configuration state to `clang` CLI options for the following reasons:
- We target one fixed platform (PS5).
- We don't attempt to support all LTO configurations, and we work closely with studios, so we are in a position to ensure that specific needs for DTLTO are met.

Ideally, for the DTLTO feature in LLVM (upstream), all LTO configurations would be supported, and the code generation would be identical to in-process ThinLTO.

Note that DTLTO needs to know:
- How to configure the tool that is used for the backend compilations to generate code correctly.
- The input/output files that will be consumed/generated by that tool for that configuration*.

*Interestingly, this is not a hard requirement for all distribution systems, as some - e.g., Incredibuild - will detect which files need transferring or returning from executions on remote agents automatically.

Thoughts and ideas:

- Add functionality to serialize/de-serialize the LTO configuration state in LLVM so that the LTO configuration state can be directly controlled.
- Perform code generation via a different tool than `clang`- perhaps `opt`? Where the mapping of CLI options to LTO configuration state is simpler?
- Add `-fthinlto-index=` support to the tool invoked to perform LTO, e.g., LLD, so that the tool could re-invoke itself with the same options it was called with + `-fthinlto-index=<blah>` to perform the code generation. This should allow the tool to set up the same LTO configuration state as when it was originally invoked to do LTO.*
- Make it the responsibility of the Clang driver to specify options to set on the invoked `clang` instances so that the code generation is appropriate and also to specify the list of additional input/output files that will be required/generated. The driver already forwards some `-plugin-opt=` options to control the code generation for ThinLTO.
- Support only a limited set of platforms and LTO configurations for now. Document this carefully and expand the support in the future as required. Allow the user to pass options to the distributor and the code generation tool to work-around and deficiencies (basically this PR's approach).

We experimented with invoking `clang -cc1` rather than `clang`. However, we concluded that it might not be scalable to translate to `-cc1` options to achieve the correct code generation; we might, in effect, end up emulating too much of the Clang driver code in maintaining this translation. There were concerns about which `cc1` options affect code generation and which sets of options would be safe to specify for any given platform. There were (possibly incorrect) concerns about this on platforms like RISC-V, which have many complexities. Having stated those worries, this may be incorrect - it would be good to get community input here into the viability of driving via `-cc1` options.

* I would like to credit Konstantin Belochapka for this intriguing idea.

teresajohnson replied:
> With or without this patch to add integrated dthin-lto, I think something like this (automatic translation of a thin-link invocation into a matching Clang backend invocation) would be useful. Optimization level and flags forwarded through Clang directly to the linker (ex: `-Wl,-mllvm,-my_codegen_option`) are those I've run into most often.
> 
> For Buck we either have this [imho unfortunate script wrapping the backend Clang invocations](https://github.com/facebook/buck2/blob/main/prelude/cxx/dist_lto/tools/dist_lto_opt_gnu.py) that translates args from the thin-link invocation to the Clang backend invocation, or we specify both sets of flags manually. @teresajohnson @MaskRay How does Bazel handle this problem?

This hasn't been too difficult for us. Generally we simply pass all of the options given to the initial clang -c invocation to the ThinLTO backend clang invocation. This covers things like -ffunction-sections and options that typically guide optimization passes and code generation nicely (since in a non-ThinLTO compile they would go to the clang -c compiler and in a distributed ThinLTO build they go to both clang invocations). Since we don't and have never supported in-process ThinLTO internally, there isn't a history of LLVM opt/code generation options going to the link step that we need to translate.

MaskRay replied:
The current distributed ThinLTO commands look like:
```
clang -c -O2 -flto=thin a.c b.c c.c
clang -flto=thin -fuse-ld=lld -Wl,--thinlto-index-only=a.rsp,--thinlto-emit-imports-files -Wl,--thinlto-prefix-replace=';lto/' elf0.o a.o -Wl,--start-lib b.o c.o -Wl,--end-lib elf1.o $opts
clang -c -O2 -fthinlto-index=lto/a.o.thinlto.bc a.o -o lto/a.o $opts
clang -c -O2 -fthinlto-index=lto/b.o.thinlto.bc b.o -o lto/b.o $opts
clang -c -O2 -fthinlto-index=lto/c.o.thinlto.bc c.o -o lto/c.o $opts
clang -fuse-ld=lld @a.rsp elf0.o elf1.o  # a.rsp contains lto/a.o and lto/b.o
```

`$opts` can be arbitrary options that affect LLVM code generation including `-mllvm -xxx=yyy`. I believe it is infeasible to enumerate all such options, though in practice supporting, say, 10 options, is able to cover 95% use cases.

To make this reliable, I believe you have to tell lld all these -f/-g/-mllvm/xxx options, perhaps in a text file (response file).

I don't think "opt" or "lld" will be a replacement for the `clang -c -O2 -fthinlto-index=lto/a.o.thinlto.bc a.o -o lto/a.o $opts` step above. opt is an internal utility (not user-facing). lld is not used for single file compilation.

Therefore, `--thinlto-remote-opt-tool` `--thinlto-remote-codegen-tool` do not feel very useful.
I believe it's fine to call it `--thinlto-remote-cc` or `--thinlto-remote-compiler`.

teresajohnson replied:
> The current distributed ThinLTO commands look like:
> 
> ```
> clang -c -O2 -flto=thin a.c b.c c.c
> clang -flto=thin -fuse-ld=lld -Wl,--thinlto-index-only=a.rsp,--thinlto-emit-imports-files -Wl,--thinlto-prefix-replace=';lto/' elf0.o a.o -Wl,--start-lib b.o c.o -Wl,--end-lib elf1.o $opts
> clang -c -O2 -fthinlto-index=lto/a.o.thinlto.bc a.o -o lto/a.o $opts
> clang -c -O2 -fthinlto-index=lto/b.o.thinlto.bc b.o -o lto/b.o $opts
> clang -c -O2 -fthinlto-index=lto/c.o.thinlto.bc c.o -o lto/c.o $opts
> clang -fuse-ld=lld @a.rsp elf0.o elf1.o  # a.rsp contains lto/a.o and lto/b.o
> ```
> 
> `$opts` can be arbitrary options that affect LLVM code generation including `-mllvm -xxx=yyy`.

That's not quite how we provide options to the current distributed ThinLTO (via bazel in google), however. There generally aren't code-generation affecting options provided to the clang thin link invocation, particularly because we never used regular LTO or in-process ThinLTO internally, where someone might start doing that. Generally IR and code gen affecting options are provided to both the pre-LTO compile and post-LTO compile actions for a module. There are mechanisms via bazel to provide them to only one of those, but generally they go to both. Typically linker options also go to both the final native link as well as the earlier ThinLTO thin link, and there are some that go specifically to the ThinLTO thin link, but typically not those that affect IR optimization and code gen.

So another option (consistent with how you would do a non-LTO compile where IR/codegen options would go to the `-c` compile step), is to somehow embed these in the bitcode. More discussion on the LLVM patch split out from this one, here: https://github.com/llvm/llvm-project/pull/127749#discussion_r1976039748

> I believe it is infeasible to enumerate all such options, though in practice supporting, say, 10 options, is able to cover 95% use cases. However, supporting `-mllvm` is really useful, which is ignored by the current approach.
> 
> To make this reliable, I believe you have to tell the distributor (through lld) all these -f/-g/-mllvm/xxx options, perhaps in a text file (response file).
> 
> Regarding commands spawned by the distributor, I don't think "opt" or "lld" will be a replacement for the `clang -c -O2 -fthinlto-index=lto/a.o.thinlto.bc a.o -o lto/a.o $opts` step above. opt is an internal utility (not user-facing). lld is not used for single file compilation.
> 
> Therefore, `--thinlto-remote-opt-tool` `--thinlto-remote-codegen-tool` do not feel very useful. I believe it's fine to call it `--thinlto-remote-cc` or `--thinlto-remote-compiler`.

bd1976bris committed:
Update clang/docs/ThinLTO.rst

Co-authored-by: Paul Kirth <paulkirth@google.com>

teresajohnson commented:
Thanks for sending this for review! I've only just skimmed the PR so far, and need to dig more into it. However, it is massive. It's nice to see how it all fits together, but for the purpose of reviewing and upstreaming, could you split it up into multiple PRs? I.e. the LLVM only changes (with your llvm-lto2 tests), then the lld changes, followed by the clang driver changes/tests?

bd1976bris commented:
> Thanks for sending this for review! I've only just skimmed the PR so far, and need to dig more into it. However, it is massive. It's nice to see how it all fits together, but for the purpose of reviewing and upstreaming, could you split it up into multiple PRs? I.e. the LLVM only changes (with your llvm-lto2 tests), then the lld changes, followed by the clang driver changes/tests?

Apologies. This was about as small as I could get the change whilst having something functional/useful. I will split up the change as suggested and open a new PR. Meanwhile, I plan to leave this open for design comments - please let me know if that is acceptable?

teresajohnson commented:
> > Thanks for sending this for review! I've only just skimmed the PR so far, and need to dig more into it. However, it is massive. It's nice to see how it all fits together, but for the purpose of reviewing and upstreaming, could you split it up into multiple PRs? I.e. the LLVM only changes (with your llvm-lto2 tests), then the lld changes, followed by the clang driver changes/tests?
> 
> Apologies. This was about as small as I could get the change whilst having something functional/useful. I will split up the change as suggested and open a new PR. Meanwhile, I plan to leave this open for design comments - please let me know if that is acceptable?

No worries and as mentioned it is good to see the full picture. Just easier to review in smaller pieces. That sounds like a great plan, thanks!

MaskRay commented on lld/test/ELF/dtlto/imports.test:
```diff
@@ -0,0 +1,69 @@
+
+## Common command line arguments. Note that the use of validate.py will cause
+## the link to fail.
+# RUN: echo "0.o 1.o \
```
Can use `#--- l.rsp`

bd1976bris replied:
I can't work out how to do this. I think I need a `RUN:` line to expand the variables.

MaskRay commented on lld/test/ELF/dtlto/imports.test:
```diff
@@ -0,0 +1,69 @@
+
+## Check that DTLTO handles imports files correctly.
+
+# RUN: rm -rf %t.dir && split-file %s %t.dir && cd %t.dir
```
could just use `%t` instead of `%t.dir` to make the line shorter...

bd1976bris replied:
Done. Thanks.

MaskRay commented on clang/test/Driver/DTLTO/dtlto.c:
```diff
@@ -0,0 +1,44 @@
+
+// REQUIRES: lld
+
+// RUN: echo "-target x86_64-linux-gnu \
```
`-target ` has been deprecated for a long time. Use `--target=` for new tests

bd1976bris replied:
Done. Thanks. Apologies, I now recall you writing about this previously.

MaskRay commented on clang/test/Driver/DTLTO/dtlto.c:
```diff
@@ -0,0 +1,44 @@
+
+/// Check the expected arguments are forwarded by default with only
+/// --thinlto-distributor=.
+// RUN: %clang -### -target x86_64-linux-gnu -fthinlto-distributor=dist.exe -fuse-ld=lld %s 2>&1 | \
```
Use `-Werror -###` and remove --implicit-check-not=warning

bd1976bris replied:
Done. Thanks.

MaskRay commented on cross-project-tests/dtlto/README.txt:
```diff
@@ -0,0 +1,2 @@
+                                                                   -*- rst -*-
```
Why not just use README.md or README.rst?

bd1976bris replied:
Yes, a `.md` makes sense. I have also added another sentence explaining why the tests are in this area.

MaskRay commented on cross-project-tests/dtlto/archive-thin.test:
```diff
@@ -0,0 +1,72 @@
+## correctly when thin archives are present.
+
+# RUN: rm -rf %t.dir && split-file %s %t.dir && cd %t.dir
+# RUN: %clang -target x86_64-linux-gnu -c foo.c -o foo.o
```
`--target=`

bd1976bris replied:
Done. Thanks.

MaskRay commented on cross-project-tests/dtlto/dtlto.c:
```diff
@@ -0,0 +1,49 @@
+// RUN:   -Xdist %llvm_src_root/utils/dtlto/local.py \
+// RUN:   --save-temps \
+// RUN:   -fuse-ld=lld \
+// RUN:   -nostdlib \
```
-nostdlib implies -nostartfiles

bd1976bris replied:
Removed `-nostartfiles`. Thanks - I have been making this mistake for years :)

MaskRay commented on cross-project-tests/dtlto/dtlto.c:
```diff
@@ -0,0 +1,49 @@
+// RUN: ls | FileCheck %s --check-prefix=LLD
+
+/// Files produced by the bitcode compilation.
+// BITCODE: dtlto.bc
```
can use `-DAG` to avoid multiple `ls | FileCheck` invocations

bd1976bris replied:
Nice. Thanks.

MaskRay commented on cross-project-tests/dtlto/dtlto.c:
```diff
@@ -0,0 +1,49 @@
+// BITCODE: dtlto.o
+
+/// The DTLTO backend emits the jobs description JSON and a summary shard.
+// BACKEND: a.{{[0-9]+}}.dist-file.json
```
`[[#]]` instead of `{{[0-9]+}}`

bd1976bris replied:
Nice. Done.

MaskRay commented on clang/include/clang/Driver/Options.td:
```diff
@@ -4082,7 +4086,9 @@ def ffinite_loops: Flag<["-"],  "ffinite-loops">, Group<f_Group>,
   Visibility<[ClangOption, CC1Option]>;
-
+def fthinlto_distributor_EQ : Joined<["-"], "fthinlto-distributor=">, Group<f_Group>,
+  HelpText<"Specifies the <path> to the distributor process executable.">, MetaVarName<"<path>">,
```
No trailing `.` (the majority convention of HelpText )

bd1976bris replied:
Thanks. Done.

MaskRay commented on clang/lib/Driver/ToolChains/Gnu.cpp:
```diff
@@ -535,6 +535,25 @@ void tools::gnutools::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+  // do LTO or not dependent upon whether there are any bitcode input files in
+  // the link.
+  if (Arg *A = Args.getLastArg(options::OPT_fthinlto_distributor_EQ)) {
+    A->claim();
```
getLastArg implies claim

bd1976bris replied:
Nice. Thanks.

MaskRay commented on lld/ELF/Config.h:
```diff
@@ -243,6 +243,8 @@ struct Config {
   llvm::SmallVector<llvm::StringRef, 0> searchPaths;
   llvm::SmallVector<llvm::StringRef, 0> symbolOrderingFile;
   llvm::SmallVector<llvm::StringRef, 0> thinLTOModulesToCompile;
+  llvm::StringRef DTLTODistributor;
```
dtltoDistributor

dtltoRemoteOptTool

bd1976bris replied:
Oops. Thanks.

MaskRay commented on lld/ELF/InputFiles.cpp:
```diff
@@ -1702,6 +1703,38 @@ static uint8_t getOsAbi(const Triple &t) {
 
+namespace dtlto {
+// Check if an archive file is a thin archive.
+bool isThinArchive(Ctx &ctx, StringRef archiveFilePath) {
```
`dtlto::isThinArchive` https://llvm.org/docs/CodingStandards.html#use-namespace-qualifiers-to-implement-previously-declared-functions

bd1976bris replied:
Thanks. However, I think that these should just be static functions now and the namespace is unnecessary. 

It would be interesting to get your ideas about this area of the code. Currently, `isThinArchive` checks the magic number at the start of the archive file to determine if it is a thin archive - that is because: 1. ELF LLD doesn't record whether an archive is thin or not and 2. I wanted to add the minimum code to LLD for now and then address this properly this once the design of DTLTO was agreed. Perhaps the archive type could be recorded in `InputFile`s?

bd1976bris replied:
These are static functions now.

MaskRay commented on lld/ELF/LTO.cpp:
```diff
@@ -186,6 +187,16 @@ BitcodeCompiler::BitcodeCompiler(Ctx &ctx) : ctx(ctx) {
         std::string(ctx.arg.thinLTOPrefixReplaceNativeObject),
         ctx.arg.thinLTOEmitImportsFiles, indexFile.get(), onIndexWrite);
+  } else if (!ctx.arg.DTLTODistributor.empty() && !ctx.bitcodeFiles.empty()) {
+    StringRef version = getenv("LLD_VERSION"); // For testing only.
```
a bit vague why this getenv is needed

bd1976bris replied:
We thought it might be useful to pass this through to the distributor (via the JSON). This could allow incompatible distributors to report a diagnostic. Whilst this isn't useful immediately it might allow better diagnostics if the implementation in LLVM changed in some incompatible way that forced the distributors to be rewritten. We don't need this in our downstream implementation, so I haven't really tested the utility. I'm happy to remove this version stuff - at least for the initial support? However, I think some sort of compatibility check may eventually be important.

bd1976bris replied:
I have removed the version stuff in this initial PR.

MaskRay commented on lld/docs/DTLTO.rst:
```diff
@@ -0,0 +1,60 @@
+
+The command line interface for DTLTO is:
+
+- `--thinlto-distributor=<path>`
```
double backticks when referring to code that should be formatted like a single backtick in markdown

bd1976bris replied:
Done. Thanks. I checked that the rendered html looked ok.

MaskRay commented on lld/test/COFF/dtlto.test:
```diff
@@ -0,0 +1,50 @@
+
+## Common command line arguments. Note that the use of validate.py will cause
+## the link to fail.
+# RUN: echo "foo.obj /entry:foo /subsystem:console \
```
Could use `#--- l.rsp`

bd1976bris replied:
I can't do this for `l.rsp` as it has macros that need expanding. However, I was able to use `#--- o.rsp`.

MaskRay reviewed (commented):
Having llvm/clang/lld changes in one make the reviewers know the full picture easily. I agree that at some point the patches shall be split.

bd1976bris committed:
Test improvements and a bug fix.

- Addressed test review comments.

- Improved content in the cross-project-tests/dtlto README.

- Reorder the code so no operations on the jobs occur until all the
  index generation tasks have completed. This fixes a intermittent
  failure that could occur if an error occurred before all the index
  generation jobs were complete. As it was intermittent it wasn't
  caught by my testing and made it into the initial PR :( In a future
  revision we will want to introduce a more robust mechanism to clean
  up the DTLTO output. For now keep the implementation simple.

bd1976bris committed:
Address non-test review comments.

bd1976bris commented:
> > Great. Actually, we wanted to add support for distributing via FASTBuild, icecream and distcc with the initial DTLTO PR. However, we investigated found that these couldn't be supported without extending the existing codebases. More details to follow..
> 
> Interesting. Interested in hearing more. We have a heavily modified version of FASTbuild and I am very comfortable in that code base - so happy to help if needed.

On second thoughts I don't want to clutter this review with a discussion about extending particular distribution systems. So, can I please ask you (and anyone else interested) to DM @romanova-ekaterina as she has been working on support for DTLTO in FASTBuild, icecream and distcc and would be happy to answer questions about it.

justincady commented:
I just want to double check my understanding. :)

If I were to convert [this example](https://gist.github.com/MaskRay/24f4e2eed208b9d8b0a3752575a665d4#distributed-thinlto) to its functional equivalent in DTLTO, would it be:

```
clang -fuse-ld=lld -O2 -flto=thin -fthinlto-distributor=distributor_process a.c b.c c.c
```

Where:

- `distributor_process` receives a JSON file specifying all intermediate jobs
    - If necessary I can use `-Xthinlto-distributor=` to pass extra arguments to `distributor_process`
- Each job invokes `--thinlto-remote-opt-tool` (defaults to clang)
- LLD waits for the jobs to complete and performs the final link

Is that correct?

bd1976bris commented:
> If I were to convert [this example](https://gist.github.com/MaskRay/24f4e2eed208b9d8b0a3752575a665d4#distributed-thinlto) to its functional equivalent in DTLTO, would it be:
> 
> ```
> clang -fuse-ld=lld -O2 -flto=thin -fthinlto-distributor=distributor_process a.c b.c c.c
> ```

In essence, yes. I was able to get the example to link using DTLTO (via the simple: [local.py](https://github.com/bd1976bris/llvm-project/blob/main/llvm/utils/dtlto/local.py)) with the following:

> clang -flto=thin -fuse-ld=lld a.c b.c c.c elf0.c elf1.c -fthinlto-distributor=$(which python3) -Xdist $LLVMSRC/llvm/utils/dtlto/local.py

This isn't a faithful port of the example because that uses --start-lib, --end-lib to make a virtual archive from `b.o` and `c.o`  but I don't think that detail matters.

> Is that correct?

Yes. A dump of the JSON generated can be seen [here](https://gist.github.com/bd1976bris/625a077aec170aace5ca8685c992f161).

bd1976bris committed:
Replace -Xdist with -Xthinlto-distributor=

bd1976bris commented:
@teresajohnson hopefully there has been enough time for design comments. I will start breaking this up into smaller commits as suggested.

thevinster commented on llvm/include/llvm/LTO/LTO.h:
```diff
@@ -294,6 +305,32 @@ ThinBackend createInProcessThinBackend(ThreadPoolStrategy Parallelism,
+/// This ThinBackend generates the index shards and then runs the individual
+/// backend jobs via an external process. It takes the same parameters as the
+/// InProcessThinBackend, however, these parameters only control the behavior
+/// when generating the index files for the modules. Addtionally:
```
typo: `Addtionally` -> `Additionally`.

thevinster reviewed (commented):
The example of using `llvm/utils/dtlto/local.py` is super useful. Could be worth mentioning that explicitly as a helpful tip for folks looking to integrate with their distribution process. Looking forward to the smaller patches.

cachemeifyoucan reviewed (commented):
Some high level comments:
* Should we just call it distributedLTO, or DLTO? Feel like we can drop the thin part for less typing, and from the user's point of view, using thinLTO infrastructure is just implementation details.
* Are you going to implement some ways to speed up archive or that will come as follow ups? During previous discussion at dev meeting, there are few suggestions for how to do it without extracting archives.

bd1976bris commented:
> Some high level comments:
> * Should we just call it distributedLTO, or DLTO? Feel like we can drop the thin part for less typing, and from the user's point of view, using thinLTO infrastructure is just implementation details.

Thanks! I'll discuss this and get back to you tomorrow. It's worth noting that both `DTLTO` and `DLTO` have the potential to be confusing - given the existence of the current "Bazel-style" (`--thinlto-index-only`) distribution for ThinLTO.

> * Are you going to implement some ways to speed up archive or that will come as follow ups? During previous discussion at dev meeting, there are few suggestions for how to do it without extracting archives.

We plan to do that as follow up work. We will eventually want to support archives. The point of this feature is to make it easy for a user to enable distribution - so we don't want restrictions such as "archives are not supported". I think it would be best to start with an RFC-like discussion thread, we will start preparing that.

Thanks to everyone that contributed ideas at the roundtable.

bd1976bris commented:
> > Some high level comments:
> > 
> > * Should we just call it distributedLTO, or DLTO? Feel like we can drop the thin part for less typing, and from the user's point of view, using thinLTO infrastructure is just implementation details.

Apologies for the slow reply. We would like to keep the "Thin"/"T" in the name. This is clearer as we are currently only going to distribute the ThinLTO part and have no plans to distribute the FullLTO part. Making the name more general could confuse users who might then expect in e.g., a mixed Thin+Full link that the Full bitcode part would also be distributed.

cachemeifyoucan commented:
> This is clearer as we are currently only going to distribute the ThinLTO part and have no plans to distribute the FullLTO part. Making the name more general could confuse users who might then expect in e.g., a mixed Thin+Full link that the Full bitcode part would also be distributed.

I don't have a strong opinion on this but I have basically the same concerns with completely opposite conclusions. To me, the distributed thinLTO makes you think there is a distributed full LTO, while just call it distributed LTO will eliminate that confusion. Distributed LTO is by nature based on thin LTO infrastructure but that doesn't need to be exposed. Isn't the LTO option to be `Full/Thin/Distributed` cleaner?

bd1976bris commented:
> I don't have a strong opinion on this but I have basically the same concerns with completely opposite conclusions. To me, the distributed thinLTO makes you think there is a distributed full LTO, while just call it distributed LTO will eliminate that confusion. Distributed LTO is by nature based on thin LTO infrastructure but that doesn't need to be exposed. 

Accepted. I think it might be worth appealing to authority here. I wonder if @MaskRay or @teresajohnson have an opinion?

> Isn't the LTO option to be `Full/Thin/Distributed` cleaner? 

Sorry, I don't entirely understand this bit, could you expand on this a bit. Are you envisioning an interface like:
clang -flto                     -> FullLTO
clang -flto=thin            -> ThinLTO
clang -flto=distributed -> DTLTO

> You can still keep `DTLTO` for `DisTributed LTO`

:)

romanova-ekaterina commented:
> > I don't have a strong opinion on this but I have basically the same concerns with completely opposite conclusions. To me, the distributed thinLTO makes you think there is a distributed full LTO, while just call it distributed LTO will eliminate that confusion. Distributed LTO is by nature based on thin LTO infrastructure but that doesn't need to be exposed.
> 
> Accepted. I think it might be worth appealing to authority here. I wonder if @MaskRay or @teresajohnson have an opinion?
> 
> > Isn't the LTO option to be `Full/Thin/Distributed` cleaner?
> 
> Sorry, I don't entirely understand this bit, could you expand on this a bit. Are you envisioning an interface like: clang -flto -> FullLTO clang -flto=thin -> ThinLTO clang -flto=distributed -> DTLTO
> 
> > You can still keep `DTLTO` for `DisTributed LTO`
> 
> :)

We prefer to keep DTLTO name. One of the reason is to distinguish our "integrated" distributed 
appoach from "non-integrated" distributed approach that has been supported for a while.
Another reason is to avoid confusion. We have been using this acronim for a while,
referring to it in RFCs and in a couple of lightning talks on developer conferences. When we had a round table last LLVM developers conference we brought up this topic about naming one more time and we all agreed to keep DTLTO name.

romanova-ekaterina commented:
> I'm fine with DTLTO as a shorthand for "integrated distributed ThinLTO".
Great! I'm glad that you support this acronim too. :)


> BTW thanks for sending the LLVM patch, I will review that tonight or more likely tomorrow.

Teresa, when reviewing, could you please focus on the design/idea rather than doing a full-fledged code review? In a day or two we will submit another PR for "no-backend" DTLTO implementation. We are doing final "touches" to this PR now. 

No-backend DLTO implementation has some important benefits/advantages. So, I guess, at this time, it will be most important to understand both designs (i.e. current implementation with DTLTO backend that Ben submitted and the alternative "no DTLTO backend" implementation that we submit a couple of from now), rather than focusing on details of implementations/nikpicks of this particular PR. 

I will try to do my best to explain the differences between both designs at the time of submission.

Hopefully, it will help us to choose the best design for using upstream or potentially do a hybrid solution, choosing the best ideas from both designs.

teresajohnson commented:
> > I'm fine with DTLTO as a shorthand for "integrated distributed ThinLTO".
> > Great! I'm glad that you support this acronim too. :)
> 
> > BTW thanks for sending the LLVM patch, I will review that tonight or more likely tomorrow.
> 
> Teresa, when reviewing, could you please focus on the design/idea rather than doing a full-fledged code review? In a day or two we will submit another PR for "no-backend" DTLTO implementation. We are doing final "touches" to this PR now.
> 
> No-backend DLTO implementation has some important benefits/advantages. So, I guess, at this time, it will be most important to understand both designs (i.e. current implementation with DTLTO backend that Ben submitted and the alternative "no DTLTO backend" implementation that we submit a couple of from now), rather than focusing on details of implementations/nikpicks of this particular PR.
> 
> I will try to do my best to explain the differences between both designs at the time of submission.
> 
> Hopefully, it will help us to choose the best design for using upstream or potentially do a hybrid solution, choosing the best ideas from both designs.

Thanks for the heads up, so I should not do a detailed code review for PR127749? Is there more info on what you mean by a "no-backend DTLTO"?

romanova-ekaterina commented:
> * Are you going to implement some ways to speed up archive or that will come as follow ups? During previous discussion at dev meeting, there are few suggestions for how to do it without extracting archives.

Definitely. 

The main idea of Integrated distributed ThinLTO approach is to make it seamless for the user to switch from using "in-process" ThinLTO to distributed ThinLTO by simply changing (or adding) one simple option on the command line. No other changes should be expected from the user in the makefiles/builds scripts. "Ease of use" will ensure acceptance of this DTLTO project by the wider audience.

So, yes, archives will be supported. 

Current PR (with DTLTO backend) doesn't support regular archives, but there will be a requirement to support them later. Adding a support for regular archive now will increase the size of the initial PR, which is already huge. Note: archive support cannot be done inside backend because module_id needs to be changed (so it could be done only prior control reaches LTO frontend). As a result, for this current PR archive support will have to be done outside DTLTO backend.

In a couple of days, we will submit an alternative PR with "no DTLTO backend" implementation. Though from the first sight it seems logical to use a separate DTLTO backend, we thought that "no DTLTO backend" implementation will be simpler, cleaner and more importantly, will leave a solid foundation for future important performance enhancements that are planning to add in the future. These performance enhancements will be impossible to implement within DTLTO backend. 

"No DTLTO backend" PR that we will present in a few days from now will have support for archives from the very beginning. The initial design will be the following: ONLY members of the archives that are part of the imports will be saved on disk. All other members will NOT be saved/unpacked. Note: this is a "poor man" solution to keep the size of the initial PR smaller. The obvious downside of this approach is the following: every link invocation will save necessary archive members on the disk, so they will be saved every time again and again after each new invocation and will not be re-used. Saved archived members will not be shared between different DTLTO processes that use the same archives.

Our downstream DTLTO implementation in Sony is different: we convert regular archives into thin archives and save them in %TEMP% directory. If the thin archive members are determined to be "not stale" they will be re-used by different DTLTO processes that use the same archives.

We will re-discuss with upstream in the future what would be the best way to handle archives. But our current thinking that it will be hybrid of these two solutions: (a) save on the disk only the members that you need (b) share unpacked archive members between different DTLTO invocation, assuming that they are not "stale". We already have a patch for it.

romanova-ekaterina commented:
> Thanks for the heads up, so I should not do a detailed code review for PR127749? Is there more info on what you mean by a "no-backend DTLTO"?

Yes, I think for now it will be better to understand the design of the current code review that Ben submitted but not to do a detailed code review just yet. In a few days ago we will propose a new PR for a different implementation for integrated DTLTO, but without the use of "DTLTO backend". That's why we call it internally "no-backend DTLTO".

After the design ideas for both PRs will be well understood, we could discuss and decide upstream which solution is better. 

I know it looks a bit strange that the same company submits two different solutions upstream. But we couldn't come to the conclusion internally, so it was decided to discuss this issue externally (upstream). 

Though it might require additional efforts to understand both designs, discussing two different solutions upstream with experts (notably you) will lead to a choosing of the best one (or eventually lead us to a hybrid of both solutions).

MaskRay commented on lld/docs/DTLTO.rst:
```diff
@@ -0,0 +1,60 @@
+  Specifies the file to execute as a distributor process.
+  If specified, ThinLTO backend compilations will be distributed.
+
+- ``--thinlto-remote-opt-tool=<path>``
```
Perhaps --thinlto-remote-compiler or --thinlto-remote-compile-tool. It's unlikely we will use the internal tool named `opt`

bd1976bris replied:
See: https://github.com/llvm/llvm-project/pull/126654#discussion_r1968738956.

bd1976bris replied:
Updated code to use `thinlto-remote-compiler`.

bd1976bris committed:
Support more than one LTO partition.

bd1976bris committed:
Sync with #127749

- Remove linker version.

- Make AddBuffer a default parameter for LTO::run() to minimize changes
  at call sites.

- Format code with clang-format, even if it deviates slightly from the
  surrounding style. I decided that CI passing for the PRs was more
  important than matching the existing style.

- Order distributor command-line options correctly.

- Remove dead code in llvm-lto2.cpp.

- Add docstrings to Python distributors.

- Update the Python mock.py distributor to validate JSON.

- Organize LLVM DTLTO tests by moving them into a dedicated "dtlto"
  directory, tidying existing tests and adding new ones.

- Implement fix for more than one LTO partition (spotted by MaskRay).

bd1976bris commented on llvm/lib/LTO/LTO.cpp:
```diff
@@ -2142,3 +2190,326 @@ std::vector<int> lto::generateModulesOrdering(ArrayRef<BitcodeModule *> R) {
+    sys::path::append(ObjFilePath, sys::path::stem(ModulePath) + "." +
+                                       itostr(Task) + "." + UID + ".native.o");
+
+    Job &J = Jobs[Task - 1]; /*Task 0 is reserved*/
```
Yes - I broke this for more than one LTO partition whilst hastily simplifying things. Thanks! Fixed now.

bd1976bris commented on llvm/lib/LTO/LTO.cpp:
```diff
@@ -2142,3 +2190,326 @@ std::vector<int> lto::generateModulesOrdering(ArrayRef<BitcodeModule *> R) {
+
+        // Common command line template.
+        JOS.attributeArray("args", [&]() {
+          JOS.value(RemoteOptTool);
```
Would you accept `thinlto-remote-codegen-tool`? This is so that the option name doesn't imply that the invoked tool is a compiler. One of the possibilities that we might want to explore is invoking LLD to do the backend compilations on the remote machines. Please see: https://github.com/llvm/llvm-project/pull/126654#discussion_r1951760716.

bd1976bris commented on clang/lib/Driver/ToolChains/Gnu.cpp:
```diff
@@ -535,6 +535,21 @@ void tools::gnutools::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+                           Twine(ToolChain.getDriver().getClangProgramPath())));
+
+    for (auto A : Args.getAllArgValues(options::OPT_Xthinlto_distributor_EQ))
+      CmdArgs.push_back(Args.MakeArgString("-mllvm=-thinlto-distributor-arg=" + A));
```
As you already worked out I used `cl::opt` options to minimize the changes to LLD ports and llvm-lto. The syntax is verbose, but LLD is usually invoked via the compiler driver so the verbose syntax is not exposed. I'm happy to use a `cl::opt` for this. However, I would like to retain a retain the LLD options  for the COFF port where LLD is often invoked directly. Does that sound ok?

bd1976bris replied:
Having looked at this again I think your suggestion of making `--thinlto-remote-opt-tool` a `cl::opt` tool makes sense. I have updated the code to match. Please take a look.

bd1976bris commented on lld/ELF/InputFiles.cpp:
```diff
@@ -1702,6 +1703,37 @@ static uint8_t getOsAbi(const Triple &t) {
+
+// Compute a thin archive member full file path.
+static std::string
+computeThinArchiveMemberFullPath(const StringRef modulePath,
```
> If two new functions can be combined. Perhaps name the new one `dtltoGetMemberPathIfThinArchive`.
> 
> ```
> if (!ctx.arg.dtltoDistributor.empty() && !archiveName.empty())
>   path = dtltoGetMemberPathIfThinArchive(...)
> ```

I believe the simplification suggestions are viable. I also think that I can remove the this->mb = mbref; line (and its accompanying extensive comment), since that was only necessary for emitting empty summary index files - which we don't need for DTLTO.

I'm also not comfortable with isThinArchive relying on checking the magic number at the start of the archive file to determine if it is a thin archive. I'm considering recording the archive type somewhere (perhaps in InputFile) so that this check becomes unnecessary. I'll test these improvements and put them up tomorrow.

>  What if it a non-thin archive? Shall we create a temporary file?

Non-thin archives with bitcode members are important for PlayStation. A reasonable implementation would be to emit a temporary file into a cache (since libraries are often external to a project and rarely modified) and use that temporary file for DTLTO. Currently, the buffer identifier assigned here is used eventually as the ModulePath stored in the summary index. However, at this point we don't know which lazy bitcode files will actually be used by the link. Therefore, we would need to either:

Write out temporary files for all bitcode members of non-thin archives now, using the temporary file path as the buffer identifier, or write out temporary files later when a lazy bitcode file is used and added to the link. If we write out later, we would need to swap the identifier for LTO or add some remapping information to the LTO link so that the temporary file path is used for loading the bitcode file instead of the identifier assigned here.

At the LLVM conference there were also suggestions that we could teach Clang and the distribution systems to handle archives natively, perhaps even understanding identifiers such as `foo.a(bar.o at 40)`.

It would be beneficial to implement a mechanism that could also be used by the existing Bazel-style distribution in case, in the future, there is a need to support bitcode members in archives for for that.

For the initial commit, however, I would prefer to support thin archives only. non-thin archive support can be added in a later commit.

bd1976bris replied:
I have simplified the code as we discussed. I also record which archives are thin when files are added to the link in `LinkerDriver::addFile` and then pass this information into the `BitcodeFile` constructor. This removes the hack where I was reopening the archive files and checking the magic bytes to determine if they were thin. The implementation is much improved now. Thanks.

MaskRay replied:
Now there are quite a few changes to  Driver and InputFiles, which actually made me more nervous. Non-dtlto use cases now incur the overhead (not that they can't, but it's not necessary to introduce these changes for the linker option feature).

These changes might turn out to be unneeded when non-thin archives are supported.

Would be nice to restore the previous version but just improve the local code and check the thin magic only when --thinlto-distributor= is specified.

bd1976bris replied:
Done. Thanks.

bd1976bris committed:
[LLD] Improve thin archive handling

- Combine new functions as suggested by MaskRay in review.

- Remove code that worked around a crash that can no longer occur now
  that the implementation is structured as a new ThinLTO backend.

- Record whether an archive was thin or not and pass this information
  into the BitcodeFile constructor. This replaces code that reopened the
  archive file and checked the archive magic.

romanova-ekaterina commented:
> Thanks for the heads up, so I should not do a detailed code review for PR127749? Is there more info on what you mean by a "no-backend DTLTO"?

"Actually, please review whatever you would like to at this point, Theresa. I don't want to get in the way of hearing what you think - we're keen to your input. I just wanted to point out that since another branch is coming, you may wish to wait until it arrives if you think a side-by-side comparison would be a good way of doing things. To clarify: that other branch won't be put up as a pull request, but we can decide how to proceed here if the general design shown in that other branch is preferred. I also mentioned that it will appear in a few days, but that's really dependent on the results of some more internal review. We're working hard on it!"

romanova-ekaterina commented:
> In a couple of days, we will submit an alternative PR with "no DTLTO backend" implementation. Though from the first sight it seems logical to use a separate DTLTO backend, we thought that "no DTLTO backend" implementation will be simpler, cleaner and more importantly, will leave a solid foundation for future important performance enhancements that are planning to add in the future. These performance enhancements will be impossible to implement within DTLTO backend.

Actually, we don't have any data to determine whether or not these ideas for performance enhancements translate into real world gains. So, let's perhaps revisit these in subsequent PRs, assuming we get something landed for starters, here.

bd1976bris committed:
Go back to previous more contained thin archive bitcode members scheme

bd1976bris committed:
Replace `remote-opt-tool` with `remote-compiler`

bd1976bris committed:
Make `thinlto-remote-compiler` a cl::opt

bd1976bris committed:
Forward Clang -mllvm options to remote compiler invocations

bd1976bris committed:
Update documentation to match recent changes

- Remove mention of the, now removed, "Linker Version" stuff.
- Remote optimization tool -> remote compiler.

bd1976bris committed:
Update python module docstrings to match the script names

bd1976bris committed:
Disable problematic Clang format for the mllvm.c test

romanova-ekaterina commented:
> Hi Reviewers! Thanks for the feedback here. I wanted to draw attention to something that I mentioned briefly in the description - the possibility of a plugin interface as opposed to invoking an external process that consumes JSON.
> 
> There are some theoretical advantages with plugins. For example, if distribution systems exist/arise that allow data to be passed back and forth between LLVM and a distribution system using memory buffers instead of files, a plugin could perhaps do that more efficiently. But we haven't done anything yet to quantify how much better this would be vs implicitly leaning on e.g. memory mapped files and the OS file cache. The distribution systems we're motivated to support from customer demand don't have such capabilities at this time.
> 
> Does anyone have any opinions on this?

The “No backend DTLTO” branch is ready. Please have a look and let us know what you think. https://github.com/romanova-ekaterina/llvm-project/pull/new/kromanova/main/integrated-DTLTO-no-backend

This comment [https://github.com/llvm/llvm-project/pull/127749#issuecomment-2727266591]

has more details about the differences between “Out of process (DTLTO) backend” branch and “No backend” DTLTO branch.

romanova-ekaterina commented:
> > Thanks for the heads up, so I should not do a detailed code review for PR127749? Is there more info on what you mean by a "no-backend DTLTO"?
> 
> Actually, please review whatever you would like to at this point, Theresa. I don't want to get in the way of hearing what you think - we're keen to your input. I just wanted to point out that since another branch is coming, you may wish to wait until it arrives if you think a side-by-side comparison would be a good way of doing things. To clarify: that other branch won't be put up as a pull request, but we can decide how to proceed here if the general design shown in that other branch is preferred. I also mentioned that it will appear in a few days, but that's really dependent on the results of some more internal review. We're working hard on it!

The “No backend DTLTO” branch is ready. Please have a look and let us know what you think. https://github.com/romanova-ekaterina/llvm-project/pull/new/kromanova/main/integrated-DTLTO-no-backend

This comment [https://github.com/llvm/llvm-project/pull/127749#issuecomment-2727266591]

has more details about the differences between “Out of process (DTLTO) backend” branch and “No backend” DTLTO branch.

romanova-ekaterina commented:
> > Thanks for the heads up, so I should not do a detailed code review for PR127749? Is there more info on what you mean by a "no-backend DTLTO"?
> 
> Actually, please review whatever you would like to at this point, Theresa. I don't want to get in the way of hearing what you think - we're keen to your input. I just wanted to point out that since another branch is coming, you may wish to wait until it arrives if you think a side-by-side comparison would be a good way of doing things. To clarify: that other branch won't be put up as a pull request, but we can decide how to proceed here if the general design shown in that other branch is preferred. I also mentioned that it will appear in a few days, but that's really dependent on the results of some more internal review. We're working hard on it!

The “No backend DTLTO” branch is ready https://github.com/romanova-ekaterina/llvm-project/pull/new/kromanova/main/integrated-DTLTO-no-backend. Please have a look and let us know what you think. This comment [[DTLTO][LLVM] Integrated Distributed ThinLTO (DTLTO) by bd1976bris · Pull Request #127749 · llvm/llvm-project](https://github.com/llvm/llvm-project/pull/127749#issuecomment-2727266591) has more details about the differences between “Out of process (DTLTO) backend” branch and “No backend” DTLTO branch.

bd1976bris commented:
I have put a PR for introducing support for Integrated Distributed ThinLTO (DTLTO) in ELF LLD. Once that's merged I will follow up with PRs for adding DTLTO support to Clang and adding DTLTO support to COFF LLD.
